source/common/grpc/google_async_client_impl.h

```diff
@@ -219,9 +219,11 @@ class GoogleAsyncStreamImpl : public RawAsyncStream,
   void sendMessageRaw(Buffer::InstancePtr&& request, bool end_stream) override;
   void closeStream() override;
   void resetStream() override;
+  // The GoogleAsyncClientImpl doesn't do Envoy watermark based flow control.
```

Comment:
As an aside, this seems bad? Is this tracked anywhere? Perhaps TODO/open issue?

Reply:
I don't think so.  @htuch this is correct or did I overlook something?  If not I'll file an issue.

Reply:
It does its own resource management, which I believe is much stricter than Envoy; it has an internal quota system and bounds resource use this way.

I don't think we should be thinking about adding watermarks in any way here, instead we should see if we can externalize some of the tunable knobs that gRPC provides to users who want to configure memory use via Envoy's config. @markdroth do you know if this is feasible?

If we go down the path of trying to meld the Envoy and Google gRPC model, we might get into the same morass we did with threading. It is possible it has some push-back mechanism for slow client/servers, @markdroth does this exist?

Reply:
There is some basic support for bounding the amount of memory that gRPC uses.  It doesn't control *all* memory use, but it is targeted at memory usage triggered by the remote server, so it will be sensitive to things like HTTP/2 flow control.  The API is here:

https://github.com/grpc/grpc/blob/a9a0739dc8acfeb46105ebc1ed1f81a1734bd801/include/grpcpp/resource_quota_impl.h

You can configure this on a channel via this call:

https://github.com/grpc/grpc/blob/311bf04ec94d7def98d80da050fd55feaabb1398/include/grpcpp/support/channel_arguments_impl.h#L90

gRPC does not have any explicit API for flow control; instead, it is handled implicitly.  When the application stops calling `Read()` on a stream, we stop releasing the flow control window, which will eventually cause the remote side to stop being able to send.

I hope this info is helpful.

Reply:
Thanks @markdroth, that is really informative. I think in this current thread, we're interested in what happens when the access logs as buffered by the gRPC client library grow too large. By default, is there some sensibly sized default resource quota?

Reply:
This sounds like resource control when using the gRPC client to proxy data. @markdroth in this case the source of data is not gRPC client - it's request logs.  As Envoy gets more requests, the logs will continue to accumulate somewhere.  Is there an API call where one can see how much memory is being consumed?  If the peer has stopped accepting data due to flow control, this client will have to buffer data as it's added, and the log source needs to know that the grpc client has gone over limit so it can start dropping logs.

Reply:
Also updating this to note it only half-fixes the associated issue.  Once it lands I can update the issue to note it only affects the google-grpc logger until we address that - better to fix half than block on the full solution.

Reply:
There is no resource quota (and therefore no hard bound for memory limit) by default.  However, if the question here is on the send side, not the receive side, then I suspect the resource quota is less of an issue, because the flow control limit there is the one on the remote peer, not the local one.

In the gRPC C++ API, there can be only one outstanding `Write()` operation on a stream at a time, and that operation will not complete until the message has cleared outgoing flow control.  That means that as long as the remote peer is not imposing flow control push-back, there will not be much memory usage on the client, and every `Write()` operation will complete ~immediately.  And when there is flow control push-back from the remote peer, the `Write()` operation will not complete until the remote peer has freed up enough flow control window for the operation to proceed.  So, in general, there will only ever be a single `Write()` call at a time buffered in gRPC on a given stream.

Reply:
Ok, sounds like everything will end up buffered in the async client impl then, at which point we could apply Envoy buffer limits there.  Thanks very much for the pointers!

Reply:
But I think though the way we use google gRPC is to post data to a processing thread, right? Do we already have code that will correctly handle backup (your new code) or is there new code needed?

Reply:
Just refamiliarised myself with this code. We only have one async write in flight, see https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/google_async_client_impl.cc#L229. The queue that grows unbounded is https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/google_async_client_impl.h#L281. See also https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/google_async_client_impl.h#L247.

I think we would need to reason about watermarks in a non-watermarked queue or do some creative use of watermarked buffers here. Either way, the problem is Envoy side, and we need to add a TODO and tracking bug to bound this queue length.

Reply:
+1 just to track in a follow up. Note also that I think the metrics service needs to be fixed in the same way so we should add a bug/follow up for that also (cc @junr03 who has been working in this area for Envoy Mobile and can likely fix this).

Reply:
The unbounded logs is still assigned to me, so I'll either doc up the details before unassigning (now I have a handle on what they are) or I'll just the google-gRPC fix as well and close it off as fixed.